### PR TITLE
ADD: negative test for non-existent flowID(get, synchronize, reroute, validate)

### PR DESCRIPTION
**Pull request by andriidovhan**
negative tests for testing a non-existent flowID were added 
 the following requests are covered:
- get
- validate
- synchronize
- validate

> **NOTE:** negative test for switch is already pushed and can be find at  https://github.com/telstra/open-kilda/commit/e6285364c8fd68b644a0c00447b6a66aed4699be

Closes: https://github.com/telstra/open-kilda/issues/1843

### CODE REVIEW

**Comment by rtretyak** [comment]
1. Please update summary to be more informative (not too detailed, but at least what kind of test is added)
2. Use keyword for closing the #1843 issue, described here: https://help.github.com/articles/closing-issues-using-keywords/

File: services/src/test-library/src/main/java/org/openkilda/testing/Constants.java [hunk 1]
```diff
@@ -27,6 +27,7 @@
     public static final Integer RULES_INSTALLATION_TIME = 5;
     public static final Integer STATS_LOGGING_TIMEOUT = 70;
     public static final SwitchId NON_EXISTENT_SWITCH_ID = new SwitchId("de:ad:be:ef:de:ad:be:ef");
+    public static final String NON_EXISTENT_FLOW_ID = "non existent flow id non existent flow id";
```
Comment (rtretyak) [inline services/src/test-library/src/main/java/org/openkilda/testing/Constants.java]: why twice? Just curious

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowValidationNegativeSpec.groovy [hunk 2]
```diff
@@ -161,4 +166,42 @@ class FlowValidationNegativeSpec extends BaseSpecification {
...
+
+    @Unroll
+    def "Unable to #action flow on a non-existent flow"() {
+        when: "Try to #action rules on a non-existent flow"
+        northbound."$method"(NON_EXISTENT_FLOW_ID)
+
+        then: "An error is received (404 code)"
+        def exc = thrown(HttpClientErrorException)
+        exc.rawStatusCode == 404
+        exc.responseBodyAsString.to(MessageError).errorMessage == "Could not reroute flow: Flow non existent flow id non existent flow id not found"
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowValidationNegativeSpec.groovy]: Please make sure that your lines are not longer than 120 characters, wrap otherwise. We do not have automatic checkstyle for groovy yet, so there'll be no red builds for this issue, but it still should be fixed.

Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowValidationNegativeSpec.groovy]: Use `NON_EXISTENT_FLOW_ID` variable directly in the error message. If that constant changes, the test should remain working correctly.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowValidationNegativeSpec.groovy [hunk 3]
```diff
@@ -161,4 +166,42 @@ class FlowValidationNegativeSpec extends BaseSpecification {
...
+
+        where:
+        action        | method
+        "synchronize" | "synchronizeFlow"
+        "reroute"     | "rerouteFlow"
+
+    }
+
+    def "Unable to get flow on a non-existent flow"() {
+        when: "Try to synchronize rules on a non-existent flow"
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowValidationNegativeSpec.groovy]: Please pay attention to step descriptions, they are important. Currently it has nothing in common with what is actually happening in the test. Same below

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowValidationNegativeSpec.groovy [hunk 4]
```diff
@@ -161,4 +166,44 @@ class FlowValidationNegativeSpec extends BaseSpecification {
...
+        then: "An error is received (404 code)"
+        def exc = thrown(HttpClientErrorException)
+        exc.rawStatusCode == 404
+        exc.responseBodyAsString.to(MessageError).errorMessage ==
+                "Can not get flow: Flow $NON_EXISTENT_FLOW_ID not found"
+    }
+
+    @Ignore("https://github.com/telstra/open-kilda/issues/2027")
+    def "Unable to validate flow on a non-existent flow"() {
+        when: "Try to validate rules on a non-existent flow"
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowValidationNegativeSpec.groovy]: `Try to validate non-existent flow`/`Try calling 'validate flow' with non-existent flow id`
1. You can validate rules 'on switch', 'on all switches of the flow', but not 'on flow' itself.
2. It will potentially validate not only rules, bot also meters
3. It creates confusion between 'switch rules validation' and 'flow validation', which are different operations

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowValidationNegativeSpec.groovy [hunk 5]
```diff
@@ -161,4 +166,44 @@ class FlowValidationNegativeSpec extends BaseSpecification {
...
+                "Could not reroute flow: Flow $NON_EXISTENT_FLOW_ID not found"
+
+        where:
+        action        | method
+        "synchronize" | "synchronizeFlow"
+        "reroute"     | "rerouteFlow"
+    }
+
+    def "Unable to get flow on a non-existent flow"() {
+        when: "Try to get rules on a non-existent flow"
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowValidationNegativeSpec.groovy]: `"Try to get rules on a non-existent flow"` would look like this:  
`getInvolvedSwitches(flow).each{ switchId -> northbound.getSwitchRules(switchId) }`   
Here you just `Trying to get flow with non-existent flow id`

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowValidationNegativeSpec.groovy [hunk 6]
```diff
@@ -161,4 +166,44 @@ class FlowValidationNegativeSpec extends BaseSpecification {
...
+        exc.responseBodyAsString.to(MessageError).errorMessage ==
+                "Could not reroute flow: Flow $NON_EXISTENT_FLOW_ID not found"
+
+        where:
+        action        | method
+        "synchronize" | "synchronizeFlow"
+        "reroute"     | "rerouteFlow"
+    }
+
+    def "Unable to get flow on a non-existent flow"() {
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowValidationNegativeSpec.groovy]: This sounds very confusing, I don't understand. can you please rephrase? Probably sth like `Unable to get flow whith non-existent flow id`

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowValidationNegativeSpec.groovy [hunk 7]
```diff
@@ -1,15 +1,20 @@
...
 import org.openkilda.functionaltests.BaseSpecification
+import org.openkilda.messaging.error.MessageError
 import org.openkilda.messaging.info.event.IslInfoData
 import org.openkilda.messaging.model.FlowDto
 import org.openkilda.model.SwitchId
 import org.openkilda.northbound.dto.flows.FlowValidationDto
 import org.openkilda.testing.model.topology.TopologyDefinition.Switch
+import org.springframework.web.client.HttpClientErrorException
+
+import static org.openkilda.testing.Constants.NON_EXISTENT_FLOW_ID
```
Comment (ylobankov) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowValidationNegativeSpec.groovy]: Import is in the wrong place. Please, install checkstyle stuff from here to fix it in the future.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowValidationNegativeSpec.groovy [hunk 8]
```diff
@@ -161,4 +166,44 @@ class FlowValidationNegativeSpec extends BaseSpecification {
         def switches = topology.getActiveSwitches()
         return [switches.first(), switches.first()]
     }
+
+    @Unroll
+    def "Unable to #action flow with non-existent flow id"() {
```
Comment (ylobankov) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowValidationNegativeSpec.groovy]: Sorry, how there can be flow with non-existent flow ID. I would rephrase it to:

Unable to #action a non-existent flow

Comment (ylobankov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowValidationNegativeSpec.groovy]: Also, it can be a single test:

```
@Unroll
def "Unable to #action flow with non-existent flow id"() {
    assumeFalse("Ignored due to issue #2027", action == "validate")
    ...
    northbound."${action}Flow"(NON_EXISTENT_FLOW_ID)
    ...

    where:
    action        | message
    "get"         | "Can not get flow: Flow $NON_EXISTENT_FLOW_ID not found"
    "synchronize" | "Could not synchronize flow: Flow $NON_EXISTENT_FLOW_ID not found"
    "reroute"     | "Could not reroute flow: Flow $NON_EXISTENT_FLOW_ID not found"
    "validate"    | "Can not validate flow: Flow $NON_EXISTENT_FLOW_ID not found"
```

File: services/src/test-library/src/main/java/org/openkilda/testing/Constants.java [hunk 9]
```diff
@@ -27,6 +29,7 @@
     public static final Integer RULES_INSTALLATION_TIME = 5;
     public static final Integer STATS_LOGGING_TIMEOUT = 70;
     public static final SwitchId NON_EXISTENT_SWITCH_ID = new SwitchId("de:ad:be:ef:de:ad:be:ef");
+    public static final String NON_EXISTENT_FLOW_ID = "fake " + UUID.randomUUID().toString();
```
Comment (ylobankov) [inline services/src/test-library/src/main/java/org/openkilda/testing/Constants.java]: I vote that 'non-existent' would be in the name. In the future, it will help us to handle logs.

Comment (rtretyak) [reply services/src/test-library/src/main/java/org/openkilda/testing/Constants.java]: Yes, `static final` should definitely not be generated different every time. I would understand if this was some kind of `NON_EXISTENT_FLOW_ID_PREFIX` and then you would add a random part every time you call it in the test, but would still find it as an overkill.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowValidationNegativeSpec.groovy [hunk 10]
```diff
@@ -161,4 +166,23 @@ class FlowValidationNegativeSpec extends BaseSpecification {
         def switches = topology.getActiveSwitches()
         return [switches.first(), switches.first()]
     }
+
+    @Unroll
+    def "Unable to #action a non-existent flow"() {
+        Assume.assumeFalse("Ignored due to issue #2027", action == "validate")
+        when: "Trying to #action rules with non-existent flow id"
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowValidationNegativeSpec.groovy]: I still don't like the description, sorry :)
Can we get rid of 'rules' word here?

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowValidationNegativeSpec.groovy [hunk 11]
```diff
@@ -161,4 +166,23 @@ class FlowValidationNegativeSpec extends BaseSpecification {
...
+    @Unroll
+    def "Unable to #action a non-existent flow"() {
+        Assume.assumeFalse("Ignored due to issue #2027", action == "validate")
+        when: "Trying to #action a non-existent flow"
+        northbound."${action}Flow"(NON_EXISTENT_FLOW_ID)
+
+        then: "An error is received (404 code)"
+        def exc = thrown(HttpClientErrorException)
+        exc.rawStatusCode == 404
+        exc.responseBodyAsString.to(MessageError).errorMessage == "$message"
```
Comment (ylobankov) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowValidationNegativeSpec.groovy]: "$message" can be just message.

"$message" -> message

File: services/src/test-library/src/main/java/org/openkilda/testing/Constants.java [hunk 12]
```diff
@@ -27,6 +29,7 @@
     public static final Integer RULES_INSTALLATION_TIME = 5;
     public static final Integer STATS_LOGGING_TIMEOUT = 70;
     public static final SwitchId NON_EXISTENT_SWITCH_ID = new SwitchId("de:ad:be:ef:de:ad:be:ef");
+    public static final String NON_EXISTENT_FLOW_ID = "non-existent " + UUID.randomUUID().toString();
```
Comment (ylobankov) [inline services/src/test-library/src/main/java/org/openkilda/testing/Constants.java]: Please, let's have the name without whitespace.

"non-existent " -> "non-existent-"
